lib/charms/vault_k8s/v0/vault_autounseal.py

```diff
@@ -0,0 +1,488 @@
+        self, relation: ops.Relation, role_id: str, secret_id: str
+    ) -> str:
+        """Create a Juju secret with the autounseal credentials.
+
+        Args:
+            relation: The relation to grant access to the secret.
+            role_id: The role id to store in the secret.
+            secret_id: The secret id to store in the secret.
+
+        Returns:
+            The secret id of the created secret.
+        """
+        secret = self.charm.app.add_secret(
+            {
+                "role-id": role_id,
+                "secret-id": secret_id,
+            },
+        )
+        secret.grant(relation)
+        if secret.id is None:
+            raise ValueError("Secret id is None")
+        return secret.id
+
+    def set_autounseal_data(
+        self,
+        relation: ops.Relation,
+        vault_address: str,
+        approle_id: str,
+        approle_secret_id: str,
+        ca_certificate: str,
+    ) -> None:
+        """Set the autounseal data in the relation databag.
+
+        Args:
+            relation: The Juju relation to set the autounseal data in.
+            vault_address: The address of the Vault server.
+            approle_id: The approle id to use when authenticating with the Vault server.
+            approle_secret_id: The approle secret id to use when authenticating with the Vault server.
+            ca_certificate: The CA certificate to use when validating the Vault server's certificate.
+        """
+        if not self.charm.unit.is_leader():
+            logger.warning(
+                "Attempting to set the autounseal data without being the leader. Ignoring the request."
+            )
+            return
+        if relation is None:
+            logger.warning("No relation found")
+            return
+        if not relation.active:
+            logger.warning("Relation is not active")
+            return
+        credentials_secret_id = self._create_autounseal_credentials_secret(
+            relation, approle_id, approle_secret_id
+        )
+        relation.data[self.charm.app].update(
+            {
+                "address": vault_address,
+                "credentials_secret_id": credentials_secret_id,
+                "ca_certificate": ca_certificate,
+            }
+        )
+
+    def get_outstanding_requests(self, relation_id: Optional[int] = None) -> List[Relation]:
+        """Get the outstanding requests for the relation."""
+        outstanding_requests: List[Relation] = []
+        requirer_requests = self.get_requirer_requests(relation_id=relation_id)
+        for relation in requirer_requests:
+            if not self._credentials_issued_for_request(relation_id=relation.id):
+                outstanding_requests.append(relation)
+        return outstanding_requests
+
+    def get_requirer_requests(self, relation_id: Optional[int] = None) -> List[Relation]:
+        """Get all requests for the relation."""
+        pending: List[Relation] = []
```

Comment:
A nit but `pending` suggests that the request is pending but it could be already fulfilled (i.e. not pending).

Reply:
nits are my favourite

yeah I think I named this when  1, it had more logic, and 2, I didn't quite follow the code I was modelling it after yet.

This code can be simplified, so I've done that instead.

Reply:
No longer relevant in https://github.com/canonical/vault-k8s-operator/pull/346/commits/aabbbfdc8b866409c01e06f79c6b288a1aef17b1